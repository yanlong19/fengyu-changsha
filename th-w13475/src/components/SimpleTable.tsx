import { defineComponent, toRefs, computed, ref, ErrorCodes } from "vue";
import { type TableProps, tableProps, type PaginationInst } from "./types";
import { formatWidth } from '../utils/tableUtils';
import Pagination from './Pagination';
import useSort from '../hooks/useSort';
import './table.css'
export default defineComponent({
  name: "SimpleTable",
  props: tableProps,
  setup(props: TableProps, { attrs, emit, slots }) {
    // 创建props的响应式引用
    const { columns, dataSource, rowKey, pagination, filter, isDefaultSort, width } = toRefs(props);

    // 通过计算属性拿到对应的展示数据
    const beforeSortData = computed(() => {
      let data: any[];
      // 用来记录传入的表格数据的深拷贝附件
      try {
        // 添加主键字段,顺便拷贝一份数据出来,暂时不考虑超过二层引用类型导致的浅拷贝问题
        data = dataSource.value.map((row: any, index: number) => (
          {
            __id: rowKey?.value ? row[rowKey.value] : index,
            ...row
          }
        ));
      } catch (error: any) {
        data = [];
        console.log('数据源格式错误,请确保是一个JSON格式的数组', error.message)
      }

      // 调用过滤
      if (filter && filter.value) {
        data = data.filter(filter.value);
      }
      return data;
    })
    const { sort, showData, getSortText } = useSort(beforeSortData, isDefaultSort, emit)
    const tableWidth = computed(() => {
      return formatWidth(width?.value || '') || '100%'
    })
    return () => {
      return (
        <div>
          <table class="my_simple_table" cellspacing="0" style={{ width: tableWidth.value }}>
            {(columns?.value || []).map(item => {
              const { width, dataIndex } = item;
              const style = {
                width: formatWidth(width || '')
              }
              return <colgroup style={style} key={dataIndex}></colgroup>
            })}
            <thead>
              <tr>
                {(columns?.value || []).map(item => (
                  <th key={item.dataIndex} class="table_head">
                    {item.title}
                    <div v-show={item.isSort} class="sort_span" onClick={() => sort(item.dataIndex)}>
                      {getSortText(item.dataIndex)}
                    </div>
                  </th>
                ))}
              </tr>
            </thead>
            <tbody>
              {showData.value.map((rowData: any) => {
                return (
                  <tr key={rowData.__id}>
                    {(columns?.value || []).map(item => <td>{rowData[item.dataIndex]}</td>)}
                  </tr>
                )
              })}
            </tbody>
          </table>
          {pagination && <Pagination {...pagination.value}></Pagination>}
        </div>
      );
    };
  },
});