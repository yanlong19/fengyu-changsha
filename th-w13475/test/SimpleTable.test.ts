import { shallowMount } from "@vue/test-utils";
import { expect, test, describe } from "vitest";
import SimpleTable from "../src/components/SimpleTable";

const columns = [{
  dataIndex: 'name',
  title: '名称',
  width: 280,
  isSort: true,
}, {
  dataIndex: 'price',
  title: '单价',
}, {
  dataIndex: 'num',
  title: '数量',
  isSort: true,
}, {
  dataIndex: 'count',
  title: '总价',
  isSort: true,
}]
const data = [
  { name: '测试数据1', price: 12.3, num: 10, count: 123 },
  { name: '测试数据22', price: 12.3, num: 10, count: 13 },
  { name: '测试数据33', price: 12.3, num: 11, count: 123 },
  { name: '测试数33', price: 12.3, num: 10, count: 123 },
  { name: '测试数据13', price: 12.3, num: 10, count: 3 }
]
const getTestData = (sortIndex?: string, sortType?: string) => {
  const data = new Array(5).fill(null).map((item, index) => ({
    name: `排序测试_${sortIndex || ''}_${sortType || ''}_${index}`, price: parseInt(Math.random() * 10000 + '', 10) / 100, num: 10, count: 123,
  }))
  return data;
}

describe('base', () => {
  // 整体最基本的渲染
  test("mount component", () => {
    const wrapper = shallowMount(SimpleTable, {
      props: {
        columns,
        dataSource: data
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  // 宽度
  test("width component", async () => {
    const wrapper = shallowMount(SimpleTable, {
      props: {
        columns,
        dataSource: data,
        width: 600
      },
    });
    expect(wrapper.find('.my_simple_table').element.style.width).toBe('600px');
    await wrapper.setProps({ width: '700' })
    expect(wrapper.find('.my_simple_table').element.style.width).toBe('700px');
    await wrapper.setProps({ width: '800px' })
    expect(wrapper.find('.my_simple_table').element.style.width).toBe('800px');
  });
  // 过滤功能测试
  test("filter component", async () => {
    const wrapper = shallowMount(SimpleTable, {
      props: {
        columns,
        dataSource: data,
        filter: (item) => {
          return item.num !== 11
        }
      },
    });
  });

  // 简单的排序测试
  // test("sort component", async () => {
  //   const wrapper = shallowMount(SimpleTable, {
  //     props: {
  //       columns,
  //       dataSource: data
  //     },
  //   });
  //   await wrapper.find('.sort_span').trigger('click');
  //   console.log(wrapper.html())
  //   expect(wrapper.find('.sort_span').element.textContent).toBe('升序');
  //   expect(wrapper.html()).toMatchSnapshot();
  // });
})





