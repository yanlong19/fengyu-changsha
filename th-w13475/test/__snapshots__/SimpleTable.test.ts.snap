// Vitest Snapshot v1

exports[`base > mount component 1`] = `
"<div>
  <table class=\\"my_simple_table\\" cellspacing=\\"0\\">
    <colgroup style=\\"width: 280px;\\"></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <thead>
      <tr>
        <th class=\\"table_head\\">名称<div class=\\"sort_span\\">无序</div>
        </th>
        <th class=\\"table_head\\">单价<div class=\\"sort_span\\" style=\\"display: none;\\">无序</div>
        </th>
        <th class=\\"table_head\\">数量<div class=\\"sort_span\\">无序</div>
        </th>
        <th class=\\"table_head\\">总价<div class=\\"sort_span\\">无序</div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>测试数据1</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据22</td>
        <td>12.3</td>
        <td>10</td>
        <td>13</td>
      </tr>
      <tr>
        <td>测试数据33</td>
        <td>12.3</td>
        <td>11</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数33</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据13</td>
        <td>12.3</td>
        <td>10</td>
        <td>3</td>
      </tr>
    </tbody>
  </table>
  <pagination-stub total=\\"0\\" pagesizeoptions=\\"10,20,40\\"></pagination-stub>
</div>"
`;

exports[`base > sort component 1`] = `
"<div>
  <table class=\\"my_simple_table\\" cellspacing=\\"0\\">
    <colgroup style=\\"width: 280px;\\"></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <thead>
      <tr>
        <th class=\\"table_head\\">名称<div class=\\"sort_span\\">升序</div>
        </th>
        <th class=\\"table_head\\">单价<div class=\\"sort_span\\" style=\\"display: none;\\">无序</div>
        </th>
        <th class=\\"table_head\\">数量<div class=\\"sort_span\\">无序</div>
        </th>
        <th class=\\"table_head\\">总价<div class=\\"sort_span\\">无序</div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>测试数33</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据1</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据13</td>
        <td>12.3</td>
        <td>10</td>
        <td>3</td>
      </tr>
      <tr>
        <td>测试数据22</td>
        <td>12.3</td>
        <td>10</td>
        <td>13</td>
      </tr>
      <tr>
        <td>测试数据33</td>
        <td>12.3</td>
        <td>11</td>
        <td>123</td>
      </tr>
    </tbody>
  </table>
  <pagination-stub total=\\"0\\" pagesizeoptions=\\"10,20,40\\"></pagination-stub>
</div>"
`;

exports[`base > width component 1`] = `
"<div>
  <table class=\\"my_simple_table\\" cellspacing=\\"0\\" style=\\"width: 600px;\\">
    <colgroup style=\\"width: 280px;\\"></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <thead>
      <tr>
        <th class=\\"table_head\\">名称<div class=\\"sort_span\\">无序</div>
        </th>
        <th class=\\"table_head\\">单价<div class=\\"sort_span\\" style=\\"display: none;\\">无序</div>
        </th>
        <th class=\\"table_head\\">数量<div class=\\"sort_span\\">无序</div>
        </th>
        <th class=\\"table_head\\">总价<div class=\\"sort_span\\">无序</div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>测试数据1</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据22</td>
        <td>12.3</td>
        <td>10</td>
        <td>13</td>
      </tr>
      <tr>
        <td>测试数据33</td>
        <td>12.3</td>
        <td>11</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数33</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据13</td>
        <td>12.3</td>
        <td>10</td>
        <td>3</td>
      </tr>
    </tbody>
  </table>
  <pagination-stub total=\\"0\\" pagesizeoptions=\\"10,20,40\\"></pagination-stub>
</div>"
`;

exports[`mount component 1`] = `
"<div>
  <table class=\\"my_simple_table\\" cellspacing=\\"0\\">
    <colgroup style=\\"width: 280px;\\"></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <thead>
      <tr>
        <th class=\\"table_head\\">名称<div class=\\"sort_span\\">无序</div>
        </th>
        <th class=\\"table_head\\">单价<div class=\\"sort_span\\" style=\\"display: none;\\">无序</div>
        </th>
        <th class=\\"table_head\\">数量<div class=\\"sort_span\\">无序</div>
        </th>
        <th class=\\"table_head\\">总价<div class=\\"sort_span\\">无序</div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>测试数据1</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据22</td>
        <td>12.3</td>
        <td>10</td>
        <td>13</td>
      </tr>
      <tr>
        <td>测试数据33</td>
        <td>12.3</td>
        <td>11</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数33</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据13</td>
        <td>12.3</td>
        <td>10</td>
        <td>3</td>
      </tr>
    </tbody>
  </table>
  <div class=\\"table_pagination\\"><span class=\\"page_span\\">上一页</span><span class=\\"page_span page_span_active\\">1</span><span class=\\"page_span\\">下一页</span><select>
      <option value=\\"5\\">5</option>
      <option value=\\"10\\">10</option>
      <option value=\\"20\\">20</option>
    </select><input><button>跳转</button></div>
</div>"
`;

exports[`sort component 1`] = `
"<div>
  <table class=\\"my_simple_table\\" cellspacing=\\"0\\">
    <colgroup style=\\"width: 280px;\\"></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <colgroup></colgroup>
    <thead>
      <tr>
        <th class=\\"table_head\\">名称<div class=\\"sort_span\\">升序</div>
        </th>
        <th class=\\"table_head\\">单价<div class=\\"sort_span\\" style=\\"display: none;\\">无序</div>
        </th>
        <th class=\\"table_head\\">数量<div class=\\"sort_span\\">无序</div>
        </th>
        <th class=\\"table_head\\">总价<div class=\\"sort_span\\">无序</div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>测试数33</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据1</td>
        <td>12.3</td>
        <td>10</td>
        <td>123</td>
      </tr>
      <tr>
        <td>测试数据13</td>
        <td>12.3</td>
        <td>10</td>
        <td>3</td>
      </tr>
      <tr>
        <td>测试数据22</td>
        <td>12.3</td>
        <td>10</td>
        <td>13</td>
      </tr>
      <tr>
        <td>测试数据33</td>
        <td>12.3</td>
        <td>11</td>
        <td>123</td>
      </tr>
    </tbody>
  </table>
  <div class=\\"table_pagination\\"><span class=\\"page_span\\">上一页</span><span class=\\"page_span page_span_active\\">1</span><span class=\\"page_span\\">下一页</span><select>
      <option value=\\"5\\">5</option>
      <option value=\\"10\\">10</option>
      <option value=\\"20\\">20</option>
    </select><input><button>跳转</button></div>
</div>"
`;
